src/Bicep.LangServer/Handlers/BicepCompletionHandler.cs

```diff
@@ -0,0 +1,126 @@
+using System.Threading.Tasks;
+using Bicep.Core;
+using Bicep.Core.SemanticModel;
+using Bicep.Core.TypeSystem;
+using Bicep.LanguageServer.CompilationManager;
+using Bicep.LanguageServer.Completions;
+using Bicep.LanguageServer.Utils;
+using OmniSharp.Extensions.LanguageServer.Protocol.Document;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Bicep.LanguageServer.Handlers
+{
+    public class BicepCompletionHandler : CompletionHandler
+    {
+        private readonly ICompilationManager compilationManager;
+
+        public BicepCompletionHandler(ICompilationManager compilationManager)
+            : base(CreateRegistrationOptions())
+        {
+            this.compilationManager = compilationManager;
+        }
+
+        public override Task<CompletionList> Handle(CompletionParams request, CancellationToken cancellationToken)
+        {
+            var completions = GetKeywordCompletions()
+                .Concat(GetSymbolCompletions(request))
+                .Concat(GetPrimitiveTypeCompletions());
+            return Task.FromResult(new CompletionList(completions, isIncomplete: false));
+        }
+
+        public override Task<CompletionItem> Handle(CompletionItem request, CancellationToken cancellationToken)
+        {
+            return Task.FromResult(request);
+        }
+
+        public override bool CanResolve(CompletionItem value)
+        {
+            return false;
+        }
+
+        private static CompletionRegistrationOptions CreateRegistrationOptions() => new CompletionRegistrationOptions
+        {
+            DocumentSelector = DocumentSelectorFactory.Create(),
+            AllCommitCharacters = new Container<string>(),
+            ResolveProvider = false,
+            TriggerCharacters = new Container<string>()
+        };
+
+        private IEnumerable<CompletionItem> GetKeywordCompletions()
+        {
+            yield return CreateKeywordCompletion(LanguageConstants.ParameterKeyword);
+            yield return CreateKeywordCompletion(LanguageConstants.VariableKeyword);
+            yield return CreateKeywordCompletion(LanguageConstants.ResourceKeyword);
+            yield return CreateKeywordCompletion(LanguageConstants.OutputKeyword);
+        }
+
+        private IEnumerable<CompletionItem> GetSymbolCompletions(CompletionParams request)
+        {
+            var context = this.compilationManager.GetCompilation(request.TextDocument.Uri);
+            if (context == null)
+            {
+                return Enumerable.Empty<CompletionItem>();
+            }
+
+            var model = context.Compilation.GetSemanticModel();
+
+            return GetAccessibleSymbols(model)
+                .Select(sym => sym.ToCompletionItem());
+        }
+
+        private IEnumerable<CompletionItem> GetPrimitiveTypeCompletions() => LanguageConstants.DeclarationTypes.Values.Select(CreateTypeCompletion);
+
+        private IEnumerable<Symbol> GetAccessibleSymbols(SemanticModel model)
```

Comment:
Next iterations will move this code into the core because it's related to how namespace lookups work during name binding. The dictionary is here to effectively remove inaccessible symbols from completion list. (Declaring a parameter called `concat` will prevent the `concat()` function from appearing in the list of completions.